docs/adr/0003-threadsafe-interfaces.md

```diff
@@ -0,0 +1,196 @@
+runtime cost: only one method call can be executed on an instance at any time. Any attempt to call an
+object method while a concurrent method is already executing, will block until the previous call has completed.
+
+The desired API for Project Nimbus includes methods that will be called synchronously from the main thread,
+and hence must not block on network or disk I/O. Such an API cannot be built with uniffi as currently
+implemented.
+
+## Decision Drivers <!-- optional -->
+
+* Enabling consumers to control the potential blocking behaviour of their generated APIs.
+* Ensure safety of the generated code.
+* Ship a solution in a timely manner to unblock Project Nimbus.
+
+## Considered Options
+
+* Option A: Do nothing, require consumers to assume that all method calls might block.
+* Option B: Let consumers mark interface definitions as threadsafe to opt in to a non-locking handlemap.
+* Option C: Insist that all interfaces be threadsafe and replace the handlemap with raw pointers.
+* Option D: Use a non-locking handlemap and rely on calling code to behave safely.
+
+## Decision Outcome
+
+Chosen option:
+ * **Option B: Let consumers mark interface definitions as threadsafe to opt in to a non-locking handlemap.**
+
+The choice here comes down to safety and simplicity. By making a more-concurrency-friendly handlemap
+we can maintain the current strict enforcement of Rust's mutability-safety and threadsafety guarantees,
+even in the face of errors in the generated bindings. It seems to be a relatively small change, and
+by making it opt-in we avoid creating busywork for other consumers who are not urgently facing this
+problem.
+
+One downside is that consumers need to opt-in to the fix, meaning that the default behavior may still
+be surprising to new consumers. We'll mitigate this with docs and will consider revisiting the default
+behaviour if the majority of consumers adopt the new approach.
+
+This choice does also punt some potential performance improvements to future work, but that seems in keeping
+with where we are in the project's lifecycle.
+
+## Pros and Cons of the Options
+
+### Option A: Do nothing, require consumers to assume that all method calls might block.
+
+Make no changes to uniffi, and instead accept the fact that method calls are executed serially.
+Document this limitation and work with consumers to update their API definitions to account for it.
+
+* Good, because we can ship this quickly.
+* Good, because we don't give up any of the safety guarantees of the current approach.
+* Bad, because it means non-trivial rework for one of our key early consumers.
+* Bad, because it makes uniffi less attractive to potential future consumers.
+* Bad, because it keeps all the run-time overhead of the handlemap.
+* Bad, because the default behaviour still has a hidden mutex, which might be a nasty surprise
+  for future consumers in the same way that it was for Nimbus.
```

Comment:
Not mentioned here is that it's a *requirement* for nimbus which would have been impossible to meet if we did nothing.